File: android/src/main/kotlin/com/pushly/pushsdk/delegates/PushSDKDelegate.kt
package com.pushly.pushsdk.delegates

import android.content.Context
import com.pushly.android.PushSDK
import com.pushly.android.PushSDK.Companion.logLevel
import com.pushly.android.enums.PNLogLevel
import com.pushly.pushsdk.FlutterResponder
import com.pushly.pushsdk.PNLogs
import com.pushly.pushsdk.PushsdkPlugin.Companion.pushSdkConfigured
import com.pushly.pushsdk.callbacks.*
import io.flutter.Log.setLogLevel
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.embedding.engine.dart.DartExecutor
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.lang.reflect.Method

class PushSDKDelegate(
    private val context: Context,
    public val channel: MethodChannel,
    private val binaryMessenger: BinaryMessenger,
): FlutterResponder(), MethodChannel.MethodCallHandler {

    companion object {
        fun create(context: Context, binaryMessenger: BinaryMessenger): PushSDKDelegate {
            return PushSDKDelegate(context, MethodChannel(binaryMessenger, "PushSDK"), binaryMessenger)
        }
    }

    init {
        initialize(context, channel, binaryMessenger)
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        PNLogs.debug("Received method call: ${call.method} with args: ${call.arguments}")
        when (call.method) {
            "setLogLevel" -> {
                val logLevel = PNLogLevel.valueOf(call.argument<String>("level")!!.uppercase())
                PNLogs.logLevel = logLevel
                PushSDK.logLevel = logLevel
                success(result)
            }
            "registerPushSDKLifecycleCallbacks" -> {
                PushSDK.registerPushSDKLifecycleCallbacks(PnPushSDKLifecycleCallbacksImpl(context, channel, binaryMessenger))
                success(result)
            }
            "registerNotificationLifecycleCallbacks" -> {
                PushSDK.registerNotificationLifecycleCallbacks(PnNotificationLifecycleCallbacksImpl(context, channel, binaryMessenger))
                success(result)
            }
            "registerPermissionLifecycleCallbacks" -> {
                PushSDK.registerPermissionLifecycleCallbacks(PnPermissionLifecycleCallbacksImpl(context, channel, binaryMessenger))
                success(result)
            }
            "setConfiguration" -> {
                PNLogs.verbose("SDK initialization $pushSdkConfigured")
                if (!pushSdkConfigured) {
                    PNLogs.verbose("Configuring Flutter PushSDK")
                    PushSDK.setConfiguration(call.argument<String>("appKey")!!, context)
                    pushSdkConfigured = true
                } else {
                    PNLogs.warn("Flutter PushSDK already initialized. Please ensure setConfiguration is only called once.")
                }
                success(result)
            }
            "showNativeNotificationPermissionPrompt" -> {
                PushSDK.showNativeNotificationPermissionPrompt(
                    completion = {granted, permission, error ->
                        invokeOnMainThread(
                            "showNativeNotificationPermissionPromptCallback",
                            mapOf(
                                "granted" to granted,
                                "status" to permission?.name,
                                "error" to error?.stackTraceToString()
                            )
                        )
                    },
                    skipConditionsEvaluation = call.argument<Boolean?>("skipConditionsEvaluation") ?: false,
                    skipFrequencyCapEvaluation = call.argument<Boolean?>("skipFrequencyCapEvaluation") ?: false,
                )
                success(result)
            }
            else -> notImplemented(result)
        }
    }
}

File: ios/Classes/Delegates/PushSDKDelegate.swift
import Pushly
import Flutter
import UIKit

internal class PushSDKDelegate: NSObject, FlutterPlugin {
    
    private let channel: FlutterMethodChannel
    private let binaryMessenger: FlutterBinaryMessenger
    
    private static var pushSdkConfigured = false
    
    static func register(with registrar: FlutterPluginRegistrar) {
        let instance = PushSDKDelegate(binaryMessenger: registrar.messenger())
        registrar.addMethodCallDelegate(instance, channel: instance.channel)
    }
    
    init(binaryMessenger: FlutterBinaryMessenger) {
        self.binaryMessenger = binaryMessenger
        channel = FlutterMethodChannel(name: "PushSDK", binaryMessenger: binaryMessenger)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        PNLogs.debug("Received method call: \(call.method) with args \(String(describing: call.arguments))")
        let args: [String: Any?] = (call.arguments as? Dictionary<String, Any?>) ?? [:]
        
        switch call.method {
        case "setLogLevel":
            let level = args["level"] as? String
            let logLevel: PNLogLevel
            switch (level?.uppercased()) {
                case "VERBOSE": logLevel = PNLogLevel.verbose; break
                case "DEBUG": logLevel = PNLogLevel.debug; break
                case "INFO": logLevel = PNLogLevel.info; break
                case "WARN": logLevel = PNLogLevel.warn; break
                case "ERROR": logLevel = PNLogLevel.error; break
                case "CRITICAL": logLevel = PNLogLevel.critical; break
                case "NONE": fallthrough
                default: logLevel = PNLogLevel.none
            }
            PNLogs.logLevel = logLevel
            PushSDK.logLevel = logLevel
            result(nil)
        case "registerPushSDKLifecycleCallbacks":
            PushSDK.setSDKLifecycleDelegate(PNPushSDKLifecycleDelegateImpl(channel: channel, binaryMessenger: binaryMessenger))
            result(nil)
        case "registerNotificationLifecycleCallbacks":
            PushSDK.setNotificationLifecycleDelegate(PushSDKNativeNotificationLifecycleDelegate(channel: channel, binaryMessenger: binaryMessenger))
            result(nil)
        case "registerPermissionLifecycleCallbacks":
            PushSDK.setPermissionsLifecycleDelegate(PNPermissionsLifecycleDelegateImpl(channel: channel, binaryMessenger: binaryMessenger))
            result(nil)
        case "setConfiguration":
            if !PushSDKDelegate.pushSdkConfigured {
                PNLogs.verbose("Configuring Flutter PushSDK")
                PushSDK.setConfiguration(appKey: args["appKey"] as! String, withLaunchOptions: nil)
                PushSDKDelegate.pushSdkConfigured = true
            } else {
                PNLogs.warn("Flutter PushSDK already initialized. Please ensure setConfiguration is only called once.")
            }
            result(nil)
        case "showNativeNotificationPermissionPrompt":
            PushSDK.showNativeNotificationPermissionPrompt({ [unowned self] granted, settings, err in
                let status = parsePermissionResponse(granted: granted, withSettings: settings)
                self.channel.invokeMethod("showNativeNotificationPermissionPromptCallback", arguments: ["granted": granted, "status": status, "error": err?.localizedDescription])
            }, skipConditionsEvaluation: args["skipConditionsEvaluation"] as! Bool, skipFrequencyCapEvaluation: args["skipFrequencyCapEvaluation"] as! Bool)
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

